### ISSUE TITLE
Need to add an API warning when 2 ContainerPorts are identical

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#118418]
It's apprently allowed to say:

```
ports:
- containerPort: 1234
  protocol: TCP
- containerPort: 1234
  protocol: TCP
```

We should issue a warning.  It's absolutely meaningless, but changing validation would potentially break people and normalizing it would potentially cause controller loops.

**Comment by thockin** [comment]
Do take a look at #113245 too

**Comment by aroradaman** [comment]
@thockin are we looking for something like this ?

```diff
index 31a66b116be..ef325382db4 100644
--- a/pkg/api/pod/warnings.go
+++ b/pkg/api/pod/warnings.go
@@ -271,6 +271,18 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
                }
        }
 
+       portProtocolSet := sets.New[string]()
+       for _, container := range podSpec.Containers {
+               for _, port := range container.Ports {
+                       // emit warning if key [<port>/<protocol>] is already defined
+                       key := fmt.Sprintf("%d-%s", port.ContainerPort, port.Protocol)
+                       if portProtocolSet.Has(key) {
+                               warnings = append(warnings, fmt.Sprintf("%d/%s: duplicate combination of port and protocol.", port.ContainerPort, port.Protocol))
+                       }
+                       portProtocolSet.Insert(key)
+               }
+       }
+
        return warnings
 }

```

**Comment by thockin** [comment]
I think we need to consider hostPort and hostIP, too, and we should accumulate across containers in a pod.

IOW: do not warn on:

```
ports:
- containerPort: 12345
  protocol: TCP
  hostPort: 123
- containerPort: 1234
  protocol: TCP
  hostPort: 345
```

...because the hostPorts are different.

Do not Warn on:

```
ports:
- containerPort: 1234
  protocol: TCP
  hostPort: 1234
  hostIP: 1.2.3.4
- containerPort: 1234
  protocol: TCP
  hostPort: 1234
 hostIP: 3.4.5.6
```

...because the ports are on different hostIPs

Do warn on:

```
ports:
- containerPort: 1234
  protocol: TCP
- containerPort: 1234
  protocol: TCP
```

...because that's totally redundant. 

Do warn on:

```
ports:
- containerPort: 1234
  protocol: TCP
  hostPort: 1234
- containerPort: 1234
  protocol: TCP
```

...because that is also redundant - the hostPort being 0 in the second is not sufficiently distinct from the first.

**Comment by pacoxu** [comment]
Just update the containerPort part to only warn when duplicate containers[*].ports with the same port, protocol, hostPort, and hostIP.

https://github.com/kubernetes/kubernetes/pull/113245

**Comment by thockin** [comment]
Fixed in #113245
